### PR TITLE
SCB-2295 Governance supports time periods of string type

### PR DESCRIPTION
**Pull request by develpoerX**
Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/SCB) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [ ] Each commit in the pull request should have a meaningful subject line and body.
 - [ ] Format the pull request title like `[SCB-XXX] Fixes bug in ApproximateQuantiles`, where you replace `SCB-XXX` with the appropriate JIRA issue.
 - [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [ ] Run `mvn clean install -Pit` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

### CODE REVIEW

File: governance/src/main/java/org/apache/servicecomb/governance/handler/RetryHandler.java [hunk 1]
```diff
@@ -65,21 +63,26 @@ protected Retry createProcessor(RetryPolicy policy) {
...
+        .intervalFunction(getIntervalFunction(retryPolicy))
         .build();
 
     RetryRegistry registry = RetryRegistry.of(config);
     return registry.retry(retryPolicy.getName());
   }
 
-  private Predicate<Object> getPredicate(List<Integer> statusList) {
+  private IntervalFunction getIntervalFunction(RetryPolicy retryPolicy) {
+    if (retryPolicy.getRetryStrategy().equals("RandomBackoff")) {
```
Comment (liubao68) [inline governance/src/main/java/org/apache/servicecomb/governance/handler/RetryHandler.java:78]: Create a String constant in RetryPolicy, and use "RandomBackoff".equls(retryPolicy.getRetryStrategy())

File: governance/src/main/java/org/apache/servicecomb/governance/handler/ext/RetryExtension.java [hunk 2]
```diff
@@ -18,9 +18,21 @@
...
 import java.util.List;
+import java.util.stream.IntStream;
 
 public interface RetryExtension {
-  boolean isRetry(List<Integer> statusList, Object result);
+  boolean isRetry(List<String> statusList, Object result);
 
   Class<? extends Throwable>[] retryExceptions();
+
+  default Boolean isContain(List<String> ststusList, String responseStatus) {
```
Comment (liubao68) [inline governance/src/main/java/org/apache/servicecomb/governance/handler/ext/RetryExtension.java:28]: It's better to create a `AbstractRetryExtension` and
1. `isContain` and `judge` give a specific name, e.g. `statusCodeConains`,  `statuceCodeMatch`
2.  give a default implementations for retry and add a default method `String extractStatusCode(Object result)`
3. make `isContain` and `judge`  private

File: governance/src/main/java/org/apache/servicecomb/governance/policy/AbstractPolicy.java [hunk 3]
```diff
@@ -27,4 +27,27 @@ public boolean isValid() {
     }
     return true;
   }
+
+  public String getTimeDuration(String time, int defaultValue) {
```
Comment (liubao68) [inline governance/src/main/java/org/apache/servicecomb/governance/policy/AbstractPolicy.java:31]: 1. add a method `Duration stringLiteratalToDuration(String time, Duration defaultValue)`
2. catch exception if time is not a valid time.

File: governance/src/test/java/org/apache/servicecomb/governance/GovernancePropertiesTest.java [hunk 4]
```diff
@@ -178,24 +178,24 @@ public void test_match_properties_changed() {
   public void test_bulkhead_properties_changed() {
     dynamicValues.put("servicecomb.bulkhead.demo-bulkhead", "rules:\n"
         + "maxConcurrentCalls: 2\n"
-        + "maxWaitDuration: 2000");
+        + "maxWaitDuration: 2000ms");
```
Comment (liubao68) [inline governance/src/test/java/org/apache/servicecomb/governance/GovernancePropertiesTest.java:181]: add new test cases, do not delete old test case for it is valid.

File: governance/src/main/java/org/apache/servicecomb/governance/handler/RetryHandler.java (hunk 1 above)
Comment (develpoerX) [reply governance/src/main/java/org/apache/servicecomb/governance/handler/RetryHandler.java:78]: OK. i will modify this. But retryPolicy.getRetryStrategy() will not be nullpoint.

File: governance/src/main/java/org/apache/servicecomb/governance/handler/ext/RetryExtension.java (hunk 2 above)
Comment (develpoerX) [reply governance/src/main/java/org/apache/servicecomb/governance/handler/ext/RetryExtension.java]: done

File: governance/src/main/java/org/apache/servicecomb/governance/policy/AbstractPolicy.java (hunk 3 above)
Comment (develpoerX) [reply governance/src/main/java/org/apache/servicecomb/governance/policy/AbstractPolicy.java]: done

File: governance/src/test/java/org/apache/servicecomb/governance/GovernancePropertiesTest.java (hunk 4 above)
Comment (develpoerX) [reply governance/src/test/java/org/apache/servicecomb/governance/GovernancePropertiesTest.java]: done

File: governance/src/main/java/org/apache/servicecomb/governance/policy/AbstractPolicy.java [hunk 5]
```diff
@@ -18,13 +18,41 @@
...
     }
     return true;
   }
+
+  private Duration parseToDuration(String time, Duration defaultValue) {
+    if (StringUtils.isEmpty(time)) {
+      return defaultValue;
+    }
+    if (time.matches(GovernanceUtils.DIGIT_REGEX)) {
+      if (Integer.valueOf(time) < 0) return Duration.ofMinutes(-1);
```
Comment (liubao68) [inline governance/src/main/java/org/apache/servicecomb/governance/policy/AbstractPolicy.java:44]: add braces

File: governance/src/main/java/org/apache/servicecomb/governance/policy/AbstractPolicy.java [hunk 6]
```diff
@@ -18,13 +18,41 @@
...
+      return Duration.ofMillis(Integer.valueOf(time));
```
Comment (liubao68) [inline governance/src/main/java/org/apache/servicecomb/governance/policy/AbstractPolicy.java:45]: Long.valueOf

File: governance/src/main/java/org/apache/servicecomb/governance/handler/ext/AbstractRetryExtension.java [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+    return isContain(statusList, statusCode);
+  }
+
+  protected abstract String extractStatusCode(Object result);
+
+  private boolean isContain(List<String> statusList, String responseStatus) {
+    return statusList.stream().anyMatch(status -> judge(status, responseStatus));
+  }
+
+  private boolean judge(String status, String responseStatus) {
```
Comment (liubao68) [inline governance/src/main/java/org/apache/servicecomb/governance/handler/ext/AbstractRetryExtension.java:40]: add some unit test cases to test status code match with '4xx'

File: handlers/handler-governance/src/main/java/org/apache/servicecomb/handler/governance/ServiceCombRetryExtension.java [hunk 8]
```diff
@@ -17,48 +17,39 @@
...
     }
-    return false;
+    return null;
   }
 
   @Override
   @SuppressWarnings({"unchecked", "rawtypes"})
   public Class<? extends Throwable>[] retryExceptions() {
-    return new Class[] {
+    return new Class[]{
```
Comment (liubao68) [inline handlers/handler-governance/src/main/java/org/apache/servicecomb/handler/governance/ServiceCombRetryExtension.java:52]: formate code

File: governance/src/main/java/org/apache/servicecomb/governance/policy/AbstractPolicy.java [hunk 9]
```diff
@@ -18,13 +18,41 @@
...
+    }
+    try {
+      return Duration.parse(GovernanceUtils.DIGIT_PREFIX + time);
+    } catch (DateTimeParseException e) {
+      LOGGER.warn("Parsed time to be a Duration failed. It will use the default value.");
+    }
+    return defaultValue;
```
Comment (liubao68) [inline governance/src/main/java/org/apache/servicecomb/governance/policy/AbstractPolicy.java:52]: should use same logic as `Integer.valueOf(time) < 0`

**Review by liubao68 — APPROVED** [review]

**Pull request merged**